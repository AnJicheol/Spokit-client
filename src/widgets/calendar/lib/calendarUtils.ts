import { getDaysInMonth, startOfMonth, getDay } from 'date-fns';

/**
 * 월요일 시작, 5주(35칸)짜리 달력 2차원 배열 반환 (빈칸은 null)
 * @param year 년도 (예: 2025)
 * @param month 월 (1~12)
 */

export function getMonthMatrix(year: number, month: number): (number | null)[][] {
  const daysInMonth = getDaysInMonth(new Date(year, month));
  const firstDay = getDay(startOfMonth(new Date(year, month)));

  const startIdx = (firstDay + 6) % 7;
  console.log('daysInMonth', daysInMonth, firstDay, startIdx);

  const cells: (number | null)[] = Array(35).fill(null);

  for (let i = 0; i < daysInMonth; ++i) {
    cells[startIdx + i] = i + 1;
  }

  return Array.from({ length: 5 }, (_, week) => cells.slice(week * 7, week * 7 + 7));
}
