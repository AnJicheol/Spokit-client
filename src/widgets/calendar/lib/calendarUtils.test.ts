import { describe, it, expect } from 'vitest';

import { getMonthMatrix } from './calendarUtils';

describe('getMonthMatrix 함수 태스트', () => {
  // 특정 월은 해당 월 - 1입니다. (ex: 7월 = 6)
  it('2025년 7월: 1일이 화요일에 위치해야 한다', () => {
    const matrix = getMonthMatrix(2025, 6);
    expect(matrix[0][0]).toBe(null);
    expect(matrix[0][1]).toBe(1);
  });

  it('2025년 7월: 31일까지 표시되어야 한다', () => {
    const matrix = getMonthMatrix(2025, 6);
    expect(matrix.flat().filter(Boolean).length).toBe(31);
    expect(matrix.flat().includes(31)).toBe(true);
  });

  it('2024년 2월(윤년): 29일까지 있어야 한다', () => {
    const matrix = getMonthMatrix(2024, 1);
    expect(matrix.flat().includes(29)).toBe(true);
    expect(matrix.flat().includes(30)).toBe(false);
  });
});
