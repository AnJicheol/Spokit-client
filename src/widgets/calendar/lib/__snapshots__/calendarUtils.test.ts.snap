// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getMonthMatrix 함수 태스트 > 전체 배열 구조를 snapshot으로 검증 1`] = `
[
  [
    null,
    null,
    null,
    null,
    1,
    2,
    3,
  ],
  [
    4,
    5,
    6,
    7,
    8,
    9,
    10,
  ],
  [
    11,
    12,
    13,
    14,
    15,
    16,
    17,
  ],
  [
    18,
    19,
    20,
    21,
    22,
    23,
    24,
  ],
  [
    25,
    26,
    27,
    28,
    29,
    30,
    31,
  ],
]
`;
